### DISCUSSION TITLE
Including .pyi files from rlib dependencies of extension modules?

### DISCUSSION THREAD

**Discussion by jmrgibson** [PyO3/maturin#1115] [Ideas]
I have a _rust library_ (rlib) that provides an abstract pyo3 interface, and other extension modules/crates that implement the interface by depending on the the rust library. The rust library contains the python API, and therefore also contains the .pyi type hints. The extension modules just provide implementations, and don't expose any new python interfaces.

I'm wondering if it's possible to have maturin automatically collect `.pyi` files from rlib dependencies?

At the moment I'm copying the rlib `.pyi` file to each extension module before invoking maturin, which works OK, but does mean it's an extra step in the build process.

**Comment by messense** [comment]
I'm not sure what you're proposing is a general use case.

> The rust library contains the python API, and therefore also contains the .pyi type hints. The extension modules just provide implementations, and don't expose any new python interfaces.

Then why not just move `.pyi` file to the extension modules?

**Reply by jmrgibson** [reply]
> Then why not just move `.pyi` file to the extension modules?

This would mean keeping multiple copies of the same `.pyi` around for every extension module. 

This isn't a huge issue for me, I'm mostly wondering if anyone else is in a similar situation, and what else they've tried.

**Reply by messense** [reply]
Perhaps a symlink could work to avoid keeping multiple copies?
